### PR TITLE
Features/token ablations

### PR DESCRIPTION
**Pull request by curt-tigges**
Added a variety of utils and experiment functions for tokenwise experiments. These are designed to be used with a HuggingFace Dataset that contains tokens, masks, and 1/0 position vectors that demarcate where instances of a given token are present.

Also includes functions for creating layerwise mean residual stream vectors for the given token.

### CODE REVIEW

File: utils/tokenwise_ablation.py [hunk 1]
```diff
@@ -0,0 +1,639 @@
...
+            freeze_ablated_logits = model(batch_tokens, return_type="logits", prepend_bos=False)
+            freeze_ablated_ld = compute_last_position_logit_diff(freeze_ablated_logits, batch_value['attention_mask'], batch_value['answers'])
+            freeze_ablated_ld_list.append(freeze_ablated_ld)
+            
+            model.reset_hooks()
+
+    return torch.cat(orig_ld_list), torch.cat(ablated_ld_list), torch.cat(freeze_ablated_ld_list)
+
+
+def compute_zeroed_attn_modified_loss(model: HookedTransformer, data_loader: DataLoader, heads_to_ablate) -> float:
```
Comment (ojh31) [inline utils/tokenwise_ablation.py:536]: split up line, type hint `heads_to_ablate`

File: utils/tokenwise_ablation.py [hunk 2]
```diff
@@ -0,0 +1,639 @@
...
+        # use the punct_pos tensor to set all positions right after punct_pos==1 to zero
+        # enter code here
+        # Step 1: Shift punct_pos tensor
+        # We use roll to shift the tensor. We pad the first column with zeros after the roll since roll is circular.
+        shifted_punct_pos = torch.roll(punct_pos, shifts=1, dims=1)
+        shifted_punct_pos[:, 0] = 0  # Set the first column to zero because roll is circular
+
+        # Step 2: Zero out loss_diff positions
+        # Use the shifted_punct_pos tensor to mask loss_diff and set those positions to zero.
+        loss_diff[shifted_punct_pos == 1] = 0
```
Comment (ojh31) [inline utils/tokenwise_ablation.py:629]: Need to discuss this in docstring

**Review by ojh31 — CHANGES_REQUESTED** [review]
Great work on this! Key issues for me:

- You make assumptions about the content of the dataloader objects which are a bit opaque. It would be great if you could use some OOP tricks to make this really clear, i.e. create subclasses of DataLoader with methods for converting between formats and detailed docstrings
- There are a few very similar methods along the lines of perform some tokenwise ablations and compute a metric. To reduce the pain of maintenance, I think it would be good to combine these where possible and use kwargs to differentiate the different use cases.

File: utils/experiments.py
[the remark this answers is not in the document]
Comment (curt-tigges) [reply utils/experiments.py:1]: Yep, deleted now

File: Dockerfile.backup
[the remark this answers is not in the document]
Comment (curt-tigges) [reply Dockerfile.backup:1]: Removed this

**Review by ojh31 — APPROVED** [review]
